js/stateless.js/src/state/utxo.ts

```diff
@@ -0,0 +1,228 @@
+import { PublicKey } from "@solana/web3.js";
+import { TlvDataElement, createTlvDataElement } from "./utxo-data";
+import { bigint254, bn, createBigint254 } from "./bigint254";
+import { BN } from "@coral-xyz/anchor";
+
+/// TODO: implement PublicKey254 type based on bigint254
+/// figure which fmt to use in indexer and on client side: since solana's regluar 'PublicKey' expects padding to 32.
+export type PublicKey254 = PublicKey;
+
+/** Describe the generic utxo details applicable to every utxo. */
+export type Utxo = {
+  /** Public key of program or user that owns the utxo */
+  owner: PublicKey;
+  /** Optional lamports attached to the utxo */
+  lamports: number | BN;
+  /** Optional data attached to the utxo */
+  data: TlvDataElement[] | null;
+  /**
+   * TODO: Implement address functionality
+   * Optional unique account ID that is persistent across transactions.
+   */
+  address?: PublicKey;
+};
+
+/** Context for utxos inserted into a state Merkle tree */
+export type MerkleContext = {
+  /** Poseidon hash of the utxo preimage  */
+  hash: bigint254;
+  /** State Merkle tree */
+  merkleTree: PublicKey;
+  /** 'hash' position within the Merkle tree */
+  leafIndex: number;
+  /** the state nullfier queue belonging to merkleTree */
+  stateNullifierQueue: PublicKey;
+};
+
+export type MerkleUpdateContext = {
+  /** Slot that the utxo was appended at */
+  slotUpdated: number;
+  /** Sequence */
+  seq: number;
+};
+
+export type MerkleContextWithMerkleProof = MerkleContext & {
+  /** Recent valid 'hash' proof path, expires after n slots */
+  merkleProof: PublicKey254[];
```

Comment:
why PublicKey254 ?

Reply:
I know they do that in compressed nfts but for zkps we need these to be big numbers not public keys

Reply:
replaced Publickey254 with BN254 (renamed from bigint254 bc there's no point in using bigint -which ATM is incompatible with anchor- as long as we're depending on BN.js indirectly. both anchor and web3js (for its PublicKey class) depend on BN.js).

i believe both are switching to native bigint in the midterm. we can reconsider then. 

we can also consider implementing a proper class for bn254 so that we can enforce the <254bit size at compile time (rn it's just a type that has the same shape as BN + runtime checkers).